### ISSUE TITLE
Bugs in Stack a shape itself

### ISSUE THREAD

**Issue by pkufool** [k2-fsa/k2#921]
I found this bug when testing `Unstack` (i.e. stacking the Unstack result should give the original RaggedShape itself).

There is something wrong when stacking one RaggedShape (actually adding an axis), especially for the shape with nested empty lists.  Here is an example to reproduce this bug,

```cpp
ContextPtr c = GetCpuContext(); // or GetCudaContext();
auto shape = RaggedShape(c, "[ [ ] [ ] [ ] [ [ [ x x x ] [ ] [ x ] [ ] [ ] ] [ ] ] [ ] ]");
std::vector<RaggedShape *> ptr;
prt.emplace_back(&shape);
auto dest = Stack(2, 1, ptr.data());
// result should be: [ [ ] [ ] [ ] [ [ [ [ x x x ] [ ] [ x ] [ ] [ ] ] ] [ ] ] [ ] ]
```

The above is just one bad case,  you can find many cases by removing line 278 in the following test code.

https://github.com/k2-fsa/k2/blob/278a0becf7feca5a73872e4dfee516dda477c2de/k2/csrc/ragged_test.cu#L256-L285

**Comment by pkufool** [comment]
I am busy doing other staff, I think it is not urgent, will fix it later, or maybe someone can help to fix it.

```
[F] /ceph-hw/kangwei/code/k2_rnnt_decode/k2/csrc/ragged.cu:416:k2::RaggedShape::Check() const::<lambda(int32_t)> Ragged shape has num_elems=2 for layers_[0],
vs. num_rows=5for the next layer.


[ Stack-Trace: ]
/ceph-hw/kangwei/code/k2_rnnt_decode/build/lib/libk2_log.so(k2::internal::GetStackTrace()+0x5b) [0x7f2189c3b82a]
./bin/cu_ragged_test(k2::internal::Logger::~Logger()+0x44) [0x55f058c3fbf2]
/ceph-hw/kangwei/code/k2_rnnt_decode/build/lib/libk2context.so(+0x3394dc) [0x7f218a3ed4dc]
/ceph-hw/kangwei/code/k2_rnnt_decode/build/lib/libk2context.so(k2::RaggedShape::Check() const+0xa51) [0x7f218a3eedd3]
./bin/cu_ragged_test(k2::RaggedShape::RaggedShape(std::vector<k2::RaggedShapeLayer, std::allocator<k2::RaggedShapeLayer> > const&, bool)+0x57) [0x55f058c411d1
]
/ceph-hw/kangwei/code/k2_rnnt_decode/build/lib/libk2context.so(k2::Stack(int, int, k2::RaggedShape**, k2::Array1<unsigned int>*)+0x73d) [0x7f218a402c75]
./bin/cu_ragged_test(+0x46054) [0x55f058c06054]
/ceph-hw/kangwei/code/k2_rnnt_decode/build/lib/libgtestd.so(void testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*
, void (testing::Test::*)(), char const*)+0x65) [0x7f212d707f53]
/ceph-hw/kangwei/code/k2_rnnt_decode/build/lib/libgtestd.so(void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, v
oid (testing::Test::*)(), char const*)+0x5a) [0x7f212d7017cf]
/ceph-hw/kangwei/code/k2_rnnt_decode/build/lib/libgtestd.so(testing::Test::Run()+0xee) [0x7f212d6df900]
/ceph-hw/kangwei/code/k2_rnnt_decode/build/lib/libgtestd.so(testing::TestInfo::Run()+0x109) [0x7f212d6e0221]
/ceph-hw/kangwei/code/k2_rnnt_decode/build/lib/libgtestd.so(testing::TestSuite::Run()+0x12c) [0x7f212d6e0928]
/ceph-hw/kangwei/code/k2_rnnt_decode/build/lib/libgtestd.so(testing::internal::UnitTestImpl::RunAllTests()+0x3a0) [0x7f212d6eb62e]
/ceph-hw/kangwei/code/k2_rnnt_decode/build/lib/libgtestd.so(bool testing::internal::HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bo
ol>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*)+0x65) [0x7f212d709181]
/ceph-hw/kangwei/code/k2_rnnt_decode/build/lib/libgtestd.so(bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>
(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*)+0x5a) [0x7f212d702789]
/ceph-hw/kangwei/code/k2_rnnt_decode/build/lib/libgtestd.so(testing::UnitTest::Run()+0xc9) [0x7f212d6ea133]
/ceph-hw/kangwei/code/k2_rnnt_decode/build/lib/libgtest_maind.so(RUN_ALL_TESTS()+0x11) [0x7f218b0a09da]
/ceph-hw/kangwei/code/k2_rnnt_decode/build/lib/libgtest_maind.so(main+0x3f) [0x7f218b0a0969]
/lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xe7) [0x7f212cf01b97]
./bin/cu_ragged_test(+0x3f8ba) [0x55f058bff8ba]

[F] /ceph-hw/kangwei/code/k2_rnnt_decode/k2/csrc/ragged.cu:541:void k2::RaggedShape::Check() const Check of RaggedShape failed, see errors printed above:Inval
id RaggedShape: {  num-axes = 5 RowSplits(1)=[ 0 0 0 0 2 2 ]cached_tot_size[1]=-1 RowSplits(2)=[ 0 1 2 3 4 5 ]RowIds(2)=[ 0 1 2 3 4 ]cached_tot_size[2]=5 RowS
plits(3)=[ 0 0 0 0 2 2 ]cached_tot_size[3]=-1 RowSplits(4)=[ 0 3 3 4 4 4 ]RowIds(4)=[ 0 0 0 2 ]cached_tot_size[4]=4 }


[ Stack-Trace: ]
/ceph-hw/kangwei/code/k2_rnnt_decode/build/lib/libk2_log.so(k2::internal::GetStackTrace()+0x5b) [0x7f2189c3b82a]
./bin/cu_ragged_test(k2::internal::Logger::~Logger()+0x44) [0x55f058c3fbf2]
/ceph-hw/kangwei/code/k2_rnnt_decode/build/lib/libk2context.so(k2::RaggedShape::Check() const+0x1a02) [0x7f218a3efd84]
./bin/cu_ragged_test(k2::RaggedShape::RaggedShape(std::vector<k2::RaggedShapeLayer, std::allocator<k2::RaggedShapeLayer> > const&, bool)+0x57) [0x55f058c411d1
]
/ceph-hw/kangwei/code/k2_rnnt_decode/build/lib/libk2context.so(k2::Stack(int, int, k2::RaggedShape**, k2::Array1<unsigned int>*)+0x73d) [0x7f218a402c75]
./bin/cu_ragged_test(+0x46054) [0x55f058c06054]
/ceph-hw/kangwei/code/k2_rnnt_decode/build/lib/libgtestd.so(void testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*
, void (testing::Test::*)(), char const*)+0x65) [0x7f212d707f53]
/ceph-hw/kangwei/code/k2_rnnt_decode/build/lib/libgtestd.so(void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, v
oid (testing::Test::*)(), char const*)+0x5a) [0x7f212d7017cf]
/ceph-hw/kangwei/code/k2_rnnt_decode/build/lib/libgtestd.so(testing::Test::Run()+0xee) [0x7f212d6df900]
/ceph-hw/kangwei/code/k2_rnnt_decode/build/lib/libgtestd.so(testing::TestInfo::Run()+0x109) [0x7f212d6e0221]
/ceph-hw/kangwei/code/k2_rnnt_decode/build/lib/libgtestd.so(testing::TestSuite::Run()+0x12c) [0x7f212d6e0928]
/ceph-hw/kangwei/code/k2_rnnt_decode/build/lib/libgtestd.so(testing::internal::UnitTestImpl::RunAllTests()+0x3a0) [0x7f212d6eb62e]
/ceph-hw/kangwei/code/k2_rnnt_decode/build/lib/libgtestd.so(bool testing::internal::HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bo
ol>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*)+0x65) [0x7f212d709181]
/ceph-hw/kangwei/code/k2_rnnt_decode/build/lib/libgtestd.so(bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>
(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*)+0x5a) [0x7f212d702789]
/ceph-hw/kangwei/code/k2_rnnt_decode/build/lib/libgtestd.so(testing::UnitTest::Run()+0xc9) [0x7f212d6ea133]
/ceph-hw/kangwei/code/k2_rnnt_decode/build/lib/libgtest_maind.so(RUN_ALL_TESTS()+0x11) [0x7f218b0a09da]
/ceph-hw/kangwei/code/k2_rnnt_decode/build/lib/libgtest_maind.so(main+0x3f) [0x7f218b0a0969]
/lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xe7) [0x7f212cf01b97]
./bin/cu_ragged_test(+0x3f8ba) [0x55f058bff8ba]
```

**Comment by pkufool** [comment]
Closed via #925
